### PR TITLE
Fix json output

### LINKED ISSUE #52: Test Output in JSON, making it difficult to read
Good afternoon all, 

I’m currently transitioning from VS Code to NeoVim, and decided to get started with LazyVim (a pre-configuration using the package manager Lazy.nvim. Apologies in advance if I have missed something obvious;

Neotest seems to complete the testing as expected (closest test, test file, test all files, etc…), I am experiencing two issues:
- the output (both in the floating window and the output panel) is formatted in JSON,
- althought the test passes, the test is shown as a fail (red cross)

See example output below:

test output floating window

For the setup, I followed the recommendations from the lazyVim website, and simply added the plugins in the relevant folder:

# neotest.lua 
[the file preconfigured by LazyVim]

```lua
return {
  {
    "folke/which-key.nvim",
    optional = true,
    opts = {
      defaults = {
        ["<leader>t"] = { name = "+test" },
      },
    },
  },
  {
    "nvim-neotest/neotest",
    opts = {
      -- Can be a list of adapters like what neotest expects,
      -- or a list of adapter names,
      -- or a table of adapter names, mapped to adapter configs.
      -- The adapter will then be automatically loaded with the config.
      adapters = {},
      -- Example for loading neotest-go with a custom config
      -- adapters = {
      --   ["neotest-go"] = {
      --     args = { "-tags=integration" },
      --   },
      -- },
      status = { virtual_text = true },
      output = { open_on_run = true },
      quickfix = {
        open = function()
          if require("lazyvim.util").has("trouble.nvim") then
            vim.cmd("Trouble quickfix")
          else
            vim.cmd("copen")
          end
        end,
      },
    },
    config = function(_, opts)
      local neotest_ns = vim.api.nvim_create_namespace("neotest")
      vim.diagnostic.config({
        virtual_text = {
          format = function(diagnostic)
            -- Replace newline and tab characters with space for more compact diagnostics
            local message = diagnostic.message:gsub("\n", " "):gsub("\t", " "):gsub("%s+", " "):gsub("^%s+", "")
            return message
          end,
        },
      }, neotest_ns)
      if opts.adapters then
        local adapters = {}
        for name, config in pairs(opts.adapters or {}) do
          if type(name) == "number" then
            if type(config) == "string" then
              config = require(config)
            end
            adapters[#adapters + 1] = config
          elseif config ~= false then
            local adapter = require(name)
            if type(config) == "table" and not vim.tbl_isempty(config) then
              local meta = getmetatable(adapter)
              if adapter.setup then
                adapter.setup(config)
              elseif meta and meta.__call then
                adapter(config)
              else
                error("Adapter " .. name .. " does not support setup")
              end
            end
            adapters[#adapters + 1] = adapter
          end
        end
        opts.adapters = adapters
      end
      require("neotest").setup(opts)
    end,
    -- stylua: ignore
    keys = {
      { "<leader>tt", function() require("neotest").run.run(vim.fn.expand("%")) end, desc = "Run File" },
      { "<leader>tT", function() require("neotest").run.run(vim.loop.cwd()) end, desc = "Run All Test Files" },
      { "<leader>tr", function() require("neotest").run.run() end, desc = "Run Nearest" },
      { "<leader>ts", function() require("neotest").summary.toggle() end, desc = "Toggle Summary" },
      { "<leader>to", function() require("neotest").output.open({ enter = true, auto_close = true }) end, desc = "Show Output" },
      { "<leader>tO", function() require("neotest").output_panel.toggle() end, desc = "Toggle Output Panel" },
      { "<leader>tS", function() require("neotest").run.stop() end, desc = "Stop" },
    },
  },
  {
    "mfussenegger/nvim-dap",
    optional = true,
    -- stylua: ignore
    keys = {
      { "<leader>td", function() require("neotest").run.run({strategy = "dap"}) end, desc = "Debug Nearest" },
    },
  },
}
```

# neotest.lua 
- Local file that I use to declare the neotest-go adapter:
```lua
return {
  "nvim-neotest/neotest",
  optional = true,
  dependencies = {
    "nvim-lua/plenary.nvim",
    "nvim-treesitter/nvim-treesitter",
    "nvim-neotest/neotest-go",
  },
  opts = {
    adapters = {
      ["neotest-go"] = {
        -- Here we can set options for neotest-go, e.g.
        -- args = { "-tags=integration" }
      },
    },
  },
}
```

# neotest-go.lua
[the file I use to declare and load neotest-go]
```lua
return {
  "nvim-neotest/neotest-go",
}
```

I don’t know what I’m missing at this point, I’ve restarted the installation of LazyVim from scratch twice, and the only thing I can think of is the adapter not interpreting or parsing the output of the test, for whatever reason,

Any help would be greatly appreciated :)

### PR DESCRIPTION
**Pull request by sergii4**
Another attempt to fix #52, #80

### CODE REVIEW

**Comment by fredrikaverpil** [comment]
@sergii4 still outputs JSON here, when having the cursor on line 9 and running "run nearest test".

Screenshot

**Comment by fredrikaverpil** [comment]
@sergii4 You did not use a three-level test, it's just two levels deep. You should try this:

```go
package main

import "testing"

func TestOdd(t *testing.T) {
	t.Run("odd", func(t *testing.T) {
		t.Run("5 is odd", func(t *testing.T) {
			if 5%2 != 1 {
				t.Error("5 is actually odd")
			}
			t.Run("7 is odd", func(t *testing.T) {
				if 7%2 != 1 {
					t.Error("7 is actually odd")
				}
			})
		})
	})
}
```
Screenshot 2024-02-24 at 14 34 33

File: neotest_go/three_level_nested_test.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+		})
+		t.Run("7 is odd", func(t *testing.T) {
+			if 7%2 != 1 {
+				t.Error("7 is actually odd")
+			}
+		})
+
+	})
+
+}
```
Comment (fredrikaverpil) [inline neotest_go/three_level_nested_test.go:20]: @sergii4 You did not use a three-level test, it's just two levels deep. You should try this:

```go
package main

import "testing"

func TestOdd(t *testing.T) {
	t.Run("odd", func(t *testing.T) {
		t.Run("5 is odd", func(t *testing.T) {
			if 5%2 != 1 {
				t.Error("5 is actually odd")
			}
			t.Run("7 is odd", func(t *testing.T) {
				if 7%2 != 1 {
					t.Error("7 is actually odd")
				}
			})
		})
	})
}
```
Screenshot 2024-02-24 at 14 34 33

Comment (fredrikaverpil) [reply neotest_go/three_level_nested_test.go:20]: @sergii4 do you think this is related?
- Issue: https://github.com/nvim-neotest/neotest-go/issues/83
- PR: https://github.com/nvim-neotest/neotest-go/pull/81

I'm thinking that you're getting a variable amount of temporary files written right now.

Comment (fredrikaverpil) [reply neotest_go/three_level_nested_test.go:20]: I just checked out @encero's branch in #81 but I still get the JSON output on third `t.Run` level.

Comment (sergii4) [reply neotest_go/three_level_nested_test.go:20]: It seems as neotest/tree-sitter upstream problem to me. I keep looking

**Comment by sergii4** [comment]
@fredrikaverpil should work now

**Comment by fredrikaverpil** [comment]
Yes!! Awesome @sergii4 🎉🎉🎉

**Review by fredrikaverpil — APPROVED** [review]
Yes!! Awesome @sergii4 🎉🎉🎉
❤️ 🚀

**Comment by fredrikaverpil** [comment]
I have a feeling this might have solved both #52 and #80. At least in my go-playground repo, I don't get the `marshal_gotest_output` errors anymore!

**Comment by sergii4** [comment]
@fredrikaverpil thanks for collaboration and examples!

**Pull request merged**